DoSomething/message_broker_producer
# Adds schema changes

Fixes #90 
- Add `pid` to `message_broker_producer_productions` table
- Add `message_broker_producer_payloads` table to support admin settings to define payload values for production entries.
- Update all existing production entries with `pid` value

**Note**: This updates needs to be done before adding code to `dosomething_mbp.install` to create the payload entries for each production. A series of functional enhancements to support DS affiliate sites using Message Broker.

**Related Milestone**:
https://github.com/DoSomething/dosomething/milestones/MB%20International%20Core

**To test**:
- with fresh copy of database from staging: `ds pull stage --db`
- run `drush updb` to trigger `message_broker_producer_update_7003`
- confirm tables exist:

```
mysql> show tables like '%message%';
+-------------------------------------+
| Tables_in_dosomething (%message%)   |
+-------------------------------------+
| message_broker_producer_payloads    |
| message_broker_producer_productions |
+-------------------------------------+
```
- confirm `plid` in `message_broker_producer_payloads` as `auto_increment` field.

```
mysql> explain message_broker_producer_payloads
    -> ;
+-------+--------------+------+-----+---------+----------------+
| Field | Type         | Null | Key | Default | Extra          |
+-------+--------------+------+-----+---------+----------------+
| plid  | int(11)      | NO   | PRI | NULL    | auto_increment |
| pid   | tinyint(4)   | NO   |     | 0       |                |
| key   | varchar(128) | NO   |     |         |                |
| value | text         | NO   |     | NULL    |                |
+-------+--------------+------+-----+---------+----------------+
```
- confirm `pid` in `message_broker_producer_productions` as `auto_increment` field.

```
mysql> explain message_broker_producer_productions
    -> ;
+--------------+------------------+------+-----+---------+----------------+
| Field        | Type             | Null | Key | Default | Extra          |
+--------------+------------------+------+-----+---------+----------------+
| machine_name | varchar(128)     | NO   |     |         |                |
| config       | text             | NO   |     | NULL    |                |
| routing_key  | varchar(128)     | NO   |     |         |                |
| status       | tinyint(4)       | NO   |     | 0       |                |
| updated      | int(10) unsigned | NO   |     | 0       |                |
| pid          | int(11)          | NO   | PRI | NULL    | auto_increment |
+--------------+------------------+------+-----+---------+----------------+
```
- confirm `pid` was populated for existing production entries.

```
mysql> select pid, machine_name from message_broker_producer_productions;
+-----+-------------------------------------+
| pid | machine_name                        |
+-----+-------------------------------------+
|   1 | transactional_campaign_group_signup |
|   2 | transactional_campaign_reportback   |
|   3 | transactional_campaign_signup       |
|   4 | transactional_user_password         |
|   5 | transactional_user_profile_update   |
|   6 | transactional_user_register         |
+-----+-------------------------------------+
```
##### The role of the `message_broker_producer_payloads` table
- All "requests" (`message_broker_producer_request()`) sent to the module to generate a message in the Message Broker system identify a "production" as defined in the `message_broker_producer_productions` table. 
- Each "request" contains a payload of values in a key/value format.
- The `message_broker_producer_payloads` table will contain key/value/production entries to define **default** payload values.
  - The values can be overridden by including a entry in the request with the same key/production values.
  - Values in the entries in the `message_broker_producer_payloads` table can contain token placeholders that will be replace with values defined in the `message_broker_producer_request()` calls. 

**An example**:

```
message_broker_producer_request(<production name>, $param['token']['first_name']);
```

where the `message_broker_producer_payloads` table contains:

```
| plid  | 1
| pid   | 6 (transactional_user_register)
| key   | body
| value | Dear !first_name , <p>Welcome to DoSomething.org</p>
```
